### DISCUSSION TITLE
Content reference block (Block Editor)

### DISCUSSION THREAD

**Discussion by jdjoshuadavison** [directus/directus#19215] [Feature Requests]
### Summary

Add a custom block to the Block Editor that allows users to insert a reference to an existing content item. Effectively, we want to be able to use the M2A interface _inside_ the Block Editor.

### Motivation

We need the ability to freely mix rich text content and references to other content items. When building content, you sometimes want to reference other content items. Currently, the only way to provide that ability is to use an M2A (page builder) field. This allows you to build your content by adding existing content items in a specific order. Unfortunately, this requires that all the “pieces” that you use to build your content must be content items of their own – it is not possible to compose some new content directly onto the current item, then reference another content item, then continue composing the content on the current item. Not only is this an uncomfortable editing experience, it means you need a content model for your “content fragments” and you’ll end up with lots of single-use content items many of which would contain just a single sentence. 
Imagine you’re building the body content for a webpage. Using the current M2A page builder method, that would look something like this:
![content-reference-example-m2a](https://github.com/directus/directus/assets/34007858/81788cae-a1c7-49f6-b338-50ae70b03d1e)
Using the proposed content reference tool within Block Editor, it would look something like this:
![content-reference-example](https://github.com/directus/directus/assets/34007858/8f6270e0-bd1b-40ef-a572-7587ea9d95c2)

### Detailed Design

When adding the Block Editor to a content model, you should first be able to choose ‘content reference’ as a tool from the list of tools to show in the toolbar. If you do enable ‘content reference’ as a tool, then you should be able to specify which collections users should be able to reference content from.
![content-reference-1-model-toolbar-crop](https://github.com/directus/directus/assets/34007858/59b57273-dd8a-4623-9b80-9345a0c62c69)

![content-reference-2-model-collections-crop](https://github.com/directus/directus/assets/34007858/5d052835-6102-4376-83e5-a2661d3abcbe)

When using the Block Editor, the user should be able to select ‘content reference’ from the toolbar.
![content-reference-3-content-toolbar-crop](https://github.com/directus/directus/assets/34007858/e37b19d0-4da9-4537-bf28-ccc46a2cf261)

This will then present 2 options ‘create item’ or ‘add existing’, which when clicked will show a dropdown list of collections.
![content-reference-5-content-create-add-options-crop](https://github.com/directus/directus/assets/34007858/16ab33af-334b-4ea7-99fb-46d7ce213789)

Selecting a collection to add existing item from will open the drawer showing all items in that collection allowing the user to select multiple items.
![content-reference-6-content-select-item-crop](https://github.com/directus/directus/assets/34007858/89026291-fb23-4108-b398-fe7fd54fc523)

Confirming the selection will take you back to the block editor, with your referenced items inserted.
![content-reference-7-content-inserted-crop](https://github.com/directus/directus/assets/34007858/d284c67a-c608-410e-8f71-b3ce98157b58)

Allowing you to continue inserting blocks and building your content around the inserted references.
![content-reference-8-content-mixed-content](https://github.com/directus/directus/assets/34007858/1bf83d4f-43fa-4be3-9210-78cdb5ef1008)

The references are stored in the JSON just like any other block editor content, that would look something like this:
```
{
    "id": "NabKe0VYG6",
    "type": "reference",
    "data": {
        "items": [
            {
                "item_id": 3,
                "collection": "Appeals"
            },
            {
                "item_id": 7,
                "collection": "Appeals"
            }
        ]
    }
}
```

In context that would look like this, for example:
```
{
    "time": 1689927130983,
    "blocks": [
        {
            "id": "WriGVyDjOl",
            "type": "header",
            "data": {
                "text": "Demo of content reference block",
                "level": 2
            },
            "tunes": {
                "alignment": {
                    "alignment": "left"
                }
            }
        },
        {
            "id": "VGgFytkydn",
            "type": "paragraph",
            "data": {
                "text": "Perhaps this is an article about two new appeals that we're launching. We might open with a paragraph introducing them and inspiring the users to want to find out more. Then we might want to insert a reference to those 2 appeals which in the front end we might render as 'cards' that show a preview of the appeal and links to the appeal page itself;"
            },
            "tunes": {
                "alignment": {
                    "alignment": "left"
                }
            }
        },
        {
            "id": "NabKe0VYG6",
            "type": "reference",
            "data": {
                "items": [
                    {
                        "item_id": 3,
                        "collection": "appeals"
                    },
                    {
                        "item_id": 7,
                        "collection": "appeals"
                    }
                ]
            }
        },
        {
            "id": "CxyKe0TPA9",
            "type": "paragraph",
            "data": {
                "text": "Then, in case the users didn't already interact with the appeals, maybe we'd have another paragraph, trying to convince them again. Finally, to finish the article off we might want to insert a divider and then provide a reference to our upcoming event."
            },
            "tunes": {
                "alignment": {
                    "alignment": "left"
                }
            }
        },
        {
            "id": "8jVBK98Cbg",
            "type": "delimiter",
            "data": {}
        },
        {
            "id": "9fg39g8HgT",
            "type": "reference",
            "data": {
                "items": [
                    {
                        "item_id": 1,
                        "collection": "events"
                    }
                ]
            }
        },
        {
            "id": "YL3zsrLCEV",
            "type": "delimiter",
            "data": {}
        },
        {
            "id": "UrzkPMIkfD",
            "type": "paragraph",
            "data": {
                "text": "Thanks for reading!"
            },
            "tunes": {
                "alignment": {
                    "alignment": "left"
                }
            }
        }
    ],
    "version": "2.27.0"
}
```

### Must Have:
- Be able to configure whether to include the content reference tool in the toolbar when adding block editor to the content model
- Be able to configure which collections items can be selected from when adding block editor to the content model
- Be able to select items to insert as a reference in the block editor

### Should Have:
- Allow inserting multiple items at the same time
- Be able to create new items as well as selecting existing items

### Could Have:
- 

### Won't Have:
- 

### Drawbacks

I tried really hard – I couldn’t think of any 👀

### Alternatives

We considered other solutions to the problem:

- Other block editors have been considered, such as TipTap editor, but this feature would still have to be custom built whatever editor is used.
- As described in ‘motivation’, we tried using the M2A Page Builder field as it is, but found it to be an insufficient editing experience.
- We considered that it should be possible for developers to add their own Editor.js plugins to the block editor, and we will be submitting that as a separate request, regardless we think this is a feature that Directus should provide out of the box.

Some of the risks of not implementing this feature are:

- The Live Preview has relieved some of the issues with the M2A approach by making it possible to see what you’re editing more easily, but the experience still feels clunky and less intuitive to use.
- Although Directus is not a CMS (it is an open data platform), being used as a CMS is one of the 3 solutions you advertise as. I believe this is a feature that users are beginning to expect from a CMS, and you may find potential customers choosing other CMS solutions due to the fact Directus doesn’t have this capability.
- The M2A approaches requires additional content models and cluttering the database with lots of short and single use records.
- Developers implement other editors (or even a custom version of Editor.js) as Directus extensions, instead of using the out of the box Block Editor interface, which could lead to:
  - The extensions marketplace becoming cluttered with a variety of very similar, but slightly different, rich text editors. This could make it harder for developers to know which extension to use and increase the risk that some extensions get abandoned as other similar ones become more successful, causing developers to have to migrate.
  - Open source developer energy is divided among many different extensions all doing a similar task, that could be united in creating extensions for the core Block Editor and improving the core product.

### Adoption Strategy

This will not be a breaking change. It will be a new tool that Directus developers can simply opt-in to by editing the content models and adding it to their block editor toolbar. Adoption will be entirely optional.

### Unresolved Questions

•	Do developers want to be able to choose whether to enable the ‘create new’ and ‘add existing’ buttons?
•	Do developers want to be able to allow/disallow duplicates?
•	Do developers want to be able to limit the number of items that can be referenced?
•	Should referenced items paginate?

**Comment by rijkvanzanten** [comment]
I think for this to work as expected, we should really implement the block editor to operate _as_ a m2a. Without doing so, we'd lose relational constraints and the ability for the api to nest data responses 🤔

**Reply by jdjoshuadavison** [reply]
Yes, that would be way better than just adding the item ID into the JSON - also a lot more complex to build, I imagine.

**Reply by rasco** [reply]
that would require each collection with an editor to have its own M2A junction collection.
could get messy setting up (and maintaining) all those collections.

and what if there are multiple editors in one collection?
i guess then the junction collection would need to reference the editor's field name as well.

**Reply by rasco** [reply]
alternative approach: there is a "directus_block_editors" system-collection which keeps track of all block editors.
-> the M2A junction could be one single bridge between this system collection and the insertable blocks.

**Comment by jdjoshuadavison** [comment]
Please be aware, the day after I submitted this feature request, [Formfcw released an extension called Flexible Editor](https://github.com/formfcw/directus-extension-flexible-editor) which largely achieves effectively want we want, but with the added benefit of creating relationships between the current item and the referenced items.

**Reply by jablonski** [reply]
I've tried the "Flexible-Editor" and it is a very good proof of concept. Setup is a little bit hard, but it is a good showcase how things should work. But in the long run I would really like to see this kind of feature in Directus Core to advertise Directus as viable CMS solution, because in the long run it would be easier to build trust with clients if they don't depend on 3rd party plugins.

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Comment by w0kyj** [comment]
Unresolved Questions:
- what do inline interfaces look like?
    - each collection get it's own block?
    - thumbnail images inline (eg reuse interface displays for the item)

**Comment by rijkvanzanten** [comment]
After giving this some more thought, I believe it makes the most sense for this core-interface to be implemented based on the many to any structure. That'll give you the flexibility for the pagination question, nested relational lookups, foreign key constraints, and helps with configuring allowed related collections, as that's already part of the regular many to any setup. I'd also think it'd be super nice if each related collection becomes it's own "Block" type inline there, so instead of having a single `reference` type with an "items" array, each collection can live mixed in with the content blocks 🤔

**Comment by formfcw** [comment]
I have created the [Flexible Editor](https://github.com/formfcw/directus-extension-flexible-editor) extension which was already mentioned earlier. While working on it, I always had the idea in mind that one day it could be integrated into the Directus core. This means everything is meant to be done the Directus way and the code should be easy to read.

In short: Flexible Editor integrates the Tiptap editor as an extension into the Directus app and utilizes an optional many-to-any (M2A) field to place and link related nodes in the editor. With [Flexible Editor Content](https://github.com/formfcw/directus-extension-flexible-editor/tree/main/content) the editors JSON content can be serialized with interactive components on the front-end. Here is a video demo:

https://github.com/directus/directus/assets/78852214/2aa3530a-a716-4da5-898f-0ec4cd28595f

## Thoughts

From this point I would like to share my thoughts and experiences in the following (_expandable_) sections:

<details><summary><strong>Editor choice</strong></summary>

Flexible Editor uses the [Tiptap](https://github.com/ueberdosis/tiptap) editor for a good reason: Inline marks like bold, italic or links are stored as JSON, while [Editor.js](https://editorjs.io) (used by Directus’ block editor) stores it as plain html (at the time of writing). The latter makes rendering these inline marks unnecessarily difficult in a component library like React or Vue. _Read more about rendering below._

Here is an example to demonstrate the difference while using the same input:

Editor.js output:

```json
{
    "time": 1696920696173,
    "blocks": [
        {
            "id": "AgITQMm-Vg",
            "type": "paragraph",
            "data": {
                "text": "How <a href=\"https://directus.io/\">inline marks</a> are stored.<br>"
            }
        }
    ],
    "version": "2.22.1"
}
```

Tiptap editor output:

```json
{
    "type": "doc",
    "content": [
        {
            "type": "paragraph",
            "content": [
                { "type": "text", "text": "How " },
                {
                    "type": "text",
                    "marks": [
                        {
                            "type": "link",
                            "attrs": {
                                "href": "https://directus.io",
                                "target": "_blank",
                                "rel": "noopener noreferrer nofollow",
                                "class": null
                            }
                        }
                    ],
                    "text": "inline marks"
                },
                { "type": "text", "text": " are stored." }
            ]
        }
    ]
}
```

---

</details>


<details><summary><strong>Setup</strong></summary>

The setup process should be simple. Since the current version of Flexible Editor is a MVP, there is still room for improvement in this regard. However, there are limitations by Directus.

Here’s what might make it feel more intuitive.

-   **All in one field.** We should not separate the JSON field from the M2A field. _Read more about it below._
-   **Relation nodes should be optional.** At the moment we are forced to select a related collection in the Relationship tab when using a field in combination with M2A. But this step should be optional, so that we can use the editor without relation nodes.
-   **Option to assign related collections to a relation node type.** When using different types of relation nodes (relation blocks, relation marks, relation inline blocks), we need to assign the related (any-)collections to one of those types. _Read more about „Types of relation nodes“ below._
-   **Option to enable selection and/or creation for each collection individually.** At the moment we are not able to do this for each individual related collection – only for all related collections of a M2A field. There are many use cases where we only want to create items of a related collection without being able to select existing ones (e.g. Gallery, Video-Embeds). There are also some use cases where we only want to select items of a related collection without being able to create new ones (e.g. Related content, Internal Links). Sometimes we want to enable both: creation and selection. But it always depends on the specific related collection. _Read more about creating/selecting relation nodes below_ – since this is only about the setup.
-   **No need to set “Item Duplication Fields”.** Duplication of relation nodes is a big deal for an editor with relation nodes. For Flexible Editor to run, we need to set these “Item Duplication Fields”. Of course, it would be better if we could skip this part. _Read more about “Duplication” below._

---

</details>


<details><summary><strong>All in one field</strong></summary>

Currently, it is not possible to combine a JSON structure and a M2A relation in one field in a meaningful way. Therefore, Flexible Editor requires two fields – a JSON field and a M2A field – which complicates the setup.

Besides the setup process, there is another aspect where the combination of JSON and M2A in one field would be useful: **no need to sync relation nodes.**

Flexible Editor is an interface for a JSON field, and uses a separate M2A-field to integrate relation nodes. This means that the relation nodes must always be kept in sync with the content of the editor. We could skip this part, if the JSON structure itself could hold all the information/values to create/update/delete the related M2A items.

For example, when adding a video-url (for embedding) via a related collection called “videos”, we are getting the following value in the M2A field …

```json
{
    "create": [
        {
            "collection": "videos",
            "item": {
                "video_url": "https://video-provider.com/2309487"
            },
            "id": "2f06f613-2348-47c9-8524-1d70831757c3"
        }
    ],
    "update": [],
    "delete": []
}
```

… and the following value in Flexible Editor (where the ID has to be the same):

```json
{
    "type": "doc",
    "content": [
        {
            "type": "relation-block",
            "attrs": {
                "id": "2f06f613-2348-47c9-8524-1d70831757c3",
                "junction": "editor_nodes",
                "collection": "videos"
            }
        }
    ]
}
```

For an “all in one field” approach it would be ideal to integrate the create-update-delete-command from above in the editors JSON structure like this:

```json
{
    "type": "doc",
    "content": [
        {
            "type": "relation-block",
            "attrs": {
                "id": "2f06f613-2348-47c9-8524-1d70831757c3",
                "collection": "videos",
                "relation": {
                    "create": [
                        {
                            "collection": "videos",
                            "item": {
                                "video_url": "https://video-provider.com/2309487"
                            },
                            "id": "2f06f613-2348-47c9-8524-1d70831757c3"
                        }
                    ],
                    "update": [],
                    "delete": []
                }
            }
        }
    ]
}
```

Then when saving the item (that holds the block editor), the create-update-delete-command(s) could be extracted and executed as before. To extract these commands of the JSON structure a callback function could be emitted inside Directus interfaces, that extracts the commands, depending on the editor implementation and its corresponding JSON structure. The latter allows extension developers to implement a different editor with a different JSON structure.

---

</details>


<details><summary><strong>Types of relation nodes</strong> (internal linking)</summary>

Flexible Editor supports the usage of …

-   **Relation Blocks**, which are references to other (collection) items
-   **Relation Marks**, where you can select a text and relate this selection to (collection) item(s) – e.g. internal links, tooltips, …
-   **Relation Inline Blocks**, which are the same as relation blocks, but could be placed inline – e.g. inline image, …

While relation inline blocks don't have that many use cases, relation marks are very important because they provide a solution for internal linking.

---

</details>


<details><summary><strong>Create and/or select items</strong></summary>

At the time of writing Flexible Editor supports creating new M2A items and inserting them into the editor. To insert existing items you need to use a [wrapper collection](https://github.com/formfcw/directus-extension-flexible-editor#tutorial-relation-node-with-existing-items). As others have mentioned, it would be ideal if we could insert multiple existing items directly and at once, as we are used to do it with the M2A Builder interface. This could be solved via the drawer by adding a create button to it.

To wrap it up, here are the three use cases (already mentioned in the setup section above):

-   **Create** opens the drawer with the create form for the related collection and inserts one relation node to the editor.
-   **Select** opens the drawer with a list of items of the related collection and inserts one or more selected relation nodes to the editor. **Note** that for „Relation marks“ (as described above), we can only select one item per selection!
-   **Create AND Select** opens the drawer with a list of items of the related collection, but provides an additional create button at the top right corner (have a look at the mock-up below), which opens/switches the drawer to another drawer with the create form for the related collection.

<img width="850" alt="drawer" src="https://github.com/directus/directus/assets/78852214/9aa2aa9e-26de-475d-98dc-b6d1ee6020ee">

This way we avoid doubling every UI button in the editor interface, like „Create News“ and „Select News“.

---

</details>


<details><summary><strong>Duplication</strong> (Set Item Duplication Fields automatically)</summary>

Duplication of related M2A items is a complex topic because there are so many use cases:

1. Select a **newly created** relation node, copy and paste (or [alt] + drag’n’drop) it
    1. into the same editor
    2. into a different editor within the same item
    3. into an editor of a different item within the same collection
    4. into an editor of an item of a different collection
    5. from an nested editor to another nested editor within the same item or collection
2. Select an **existing** relation node, copy and paste (or [alt] + drag’n’drop) it
    1. into the same editor
    2. into a different editor within the same item
    3. into an editor of a different item within the same collection
    4. into an editor of an item of a different collection
    5. from an nested editor to another nested editor within the same item or collection
3. Select an existing relation node **with edits**, copy and paste (or [alt] + drag’n’drop) it
    1. into the same editor
    2. into a different editor within the same item
    3. into an editor of a different item within the same collection
    4. into an editor of an item of a different collection
    5. from an nested editor to another nested editor within the same item or collection
4. Duplicate the editor with its relation nodes together with the entire item via **Save as copy**
5. Duplicate the editor **including edits** on its relation nodes together with the entire item via **Save as copy**
6. …

Apart from all these use cases, we would have a bad experience when we had to specify the “Item Duplication Fields” for each related collection and for each editor field we use. There is no question that we need these rules, but I think we could provide a default behaviour for this. Have a look at this [feature request about “Set Item Duplication Fields automatically”](https://github.com/directus/directus/discussions/20024).

---

</details>


<details><summary><strong>Fetching data</strong></summary>

At the time of writing, Flexible Editor needs to retrieve the editors JSON content and its relation nodes separately, like …

```js
() => [
    // Fetch the editors JSON content
    "description",
    // Fetch the editors relation nodes
    "relation_nodes.gallery.*",
    "relation_nodes.video.*",
    "relation_nodes.related_content.*",
];
```

… and then merge it on the front-end like:

```js
import { injectDataIntoContent } from "directus-extension-flexible-editor/content";
const data = await fetchDirectusData();
injectDataIntoContent(data.editor_nodes, data.description);
```

It would be nice to skip the last part and inject the data already on the backend like:

```js
() => [
    "description.gallery.*",
    "description.video.*",
    "description.related_content.*",
];
```

But I think if we don't use two separate fields (read more above at “All in one field”), this would happen anyway.

---

</details>


<details><summary><strong>Rendering in the front-end</strong> (SDK)</summary>

When it comes to rendering the JSON Content generated by the editor, we need to provide a way to serialize it with interactive components, aside from just [rendering plain HTML or plain text](https://github.com/formfcw/directus-extension-flexible-editor#rendering-plain-html-or-plain-text).

Flexible Editor provides [Flexible Editor Content](https://github.com/formfcw/directus-extension-flexible-editor/tree/main/content), where you can pass the data to (Vue) components, so that you can fully customize the output for each individual node (block or mark). For example, you could take a link node and render its data using your own link component – or a router link component provided by your framework. But the main focus here is to serialize relation nodes. Read more about this in [the Flexible Editor Content docs](https://github.com/formfcw/directus-extension-flexible-editor/tree/main/content#vue) or check out this [example on StackBlitz](https://stackblitz.com/github/formfcw/directus-extension-flexible-editor/tree/main/content/vue/example-app?file=src%2Fcontent.json,src%2FApp.vue).

Such a toolset could be provided through the SDK and offer components for the most common frameworks like Vue, React, and so on.

</details>


---

As you can see, there are many aspects associated with an “editor with content references”. I've put a lot of work into this project and would be happy if it could be integrated into the Directus core, as I am still interested in pushing this “beast” forward.

**Reply by JohannesPertl** [reply]
Any update? Including this in core would be amazing

**Reply by jeremyzilar** [reply]
Great work here @formfcw ✨ 
I would also consider adding in comments around how to best import HTML data into a `flexible editor` field using the SDK. I have been working on a migration of a very large site into `flexible_editor`, and there are a few road bumps when moving from a rich-text field with HTML into a structured JSON format.

**Reply by jeremyzilar** [reply]
Let me know how I can help !

**Reply by br41nslug** [reply]
> I would also consider adding in comments around how to best import HTML data into a flexible editor field using the SDK.

How do you mean import HTML data? migrating from a wysiwyg to structured json is very custom to your wishes/project as your frontend will need to render the JSON into the desired html instead of just getting a HTML string so there will not be a 1 to 1 translation there.

> into a flexible editor field using the SDK

For the SDK the content of the flexible editor is just a JSON blob so the transformation will have to be done before handing the final JSON to the SDK to submit it.

**Reply by jeremyzilar** [reply]
It would be nice to document the shape of the JSON blob that the Flexible Editor accepts. 

I am working on the JSON APIs that represent the HTML content in our current wysiwyg. One of the things that took a bit to figure out was how to write an HTML parser that got the correct shape of the JSON and the types. For example, I learned that `<p>` tags need to be `"type": "paragraph"`, not `"type": "p"`. Now I am onto figuring out the correct types for `div`, `cite`, `blockquote`, etc...

**Reply by jeremyzilar** [reply]
Oh now I see that these managed by the TipTap editor https://tiptap.dev/docs/editor/api/nodes

**Reply by willemmulder** [reply]
@formfcw great work on the plugin! :-)

**Comment by jeremyzilar** [comment]
Something else to consider — Make the relationship happen at the record level first. Then make it possible to choose which of the related items appear in the content body.

### Example:
Let's say I have a publication that produces a lot of articles, and also a lot of videos. In the article that I am writing, I am referencing in the text a number of the videos we produce, but I really only want to embed one on the page and plan on linking off to the rest. Here is how I produce that:
- add all of the related videos via a M2M field that I have set up called "Videos". This expresses in the data that there is a significant relationship between all of these videos and the article I am writing
- next add the Video I want to include in the content body via a shortcode `[video 567]` or full video URL using `oEmbed`.

This reduces the responsibility that the content body field has to manage all of these relationships, both technically, and for editors. It also opens up options in that these content relationships are part final JSON output.

**Reply by formfcw** [reply]
For your use case, I would suggest creating two separate relationships like `video_links` and `video_embeds`. Both relate to your video collection. On the Frontend you render them differently.

Or even better: Create one relationship `videos` and add an option `Embed video`/`Display link` to the collection, where users can decide how to display the video.

I won't recommend manually referencing cryptic video codes inside a text editor.

**Reply by snimavat** [reply]
@formfcw  Can u give more details about how do u use shortcodes ? do you process them on frontend?

**Reply by formfcw** [reply]
@snimavat I don’t use shortcodes.

**Comment by maazamaani** [comment]
No more updates?
To create a page builder data model made up custom blocks, best solution today is to use M2A relation. but display is not as good as block editor.. so if there was an interface for M2A with this features in mind, that was a good choice:
- custom ui for each block or sth like visual editor (currently visual editor is not usable for page builder, because ot static url routes)
- rtl support

**Reply by formfcw** [reply]
Could you please tell us more about this:

> currently visual editor is not usable for page builder, because ot static url routes

**Reply by maazamaani** [reply]
> Could you please tell us more about this:
> 
> > currently visual editor is not usable for page builder, because ot static url routes

Imagine you have created a page builder in directus using m2a relation.. currently in visual editor you need to specify urls manually, so in a scenario that page are dynamically building, everytime you need to add url in visual editor.. so its not usable like elementor..
Live preview also has a dual screen ui and fails to create that good experience.. but a visual m2a interface or a visual editor instead of live preview inside edit page, could create that..

**Reply by veob** [reply]
@maazamaani in case it's helpful, there is some UI for M2A available in this extension (recommended by Directus team). It is pretty basic, but still more visual than native M2A: https://github.com/directus-labs/extensions/tree/main/packages/experimental-m2a-interface.

**Reply by maazamaani** [reply]
@veob
thanks for your suggestion. I've seen this interface that is actively developing: https://github.com/smartlabsAT/directus-expandable-blocks?tab=readme-ov-file
good to see this in directus core, however it's not as good as expected while its not reflecting real result.. code block makes more sense for creating medium like experience..

**Comment by JamesW1** [comment]
Thanks for the suggestion! We've moved feature requests over to our public roadmap at https://roadmap.directus.io/roadmap

I've migrated this request there for you: https://roadmap.directus.io/roadmap?id=539145a7-5ccc-4f4a-9308-a9a651c93dcf
